File: contract/contract.go
package contract

import (
	"bytes"
	"crypto/sha256"
	"encoding/json"
	"fmt"
	"io"
	"regexp"
	"sort"
	"strings"
)

const CurrentIRVersion = "1.0"

type Contract struct {
	IRVersion string   `json:"ir_version"`
	Metadata  Metadata `json:"contract"`
	Input     Input    `json:"input"`
	Rules     []Rule   `json:"rules"`
}

type Metadata struct {
	ID      string `json:"id"`
	Version string `json:"version"`
}

type Input struct {
	Kind string `json:"kind"`
}

type Rule struct {
	ID        string    `json:"id"`
	Version   string    `json:"version,omitempty"`
	Kind      string    `json:"kind"`
	Path      string    `json:"path"`
	Predicate Predicate `json:"predicate"`
	OnBreach  Policy    `json:"on_breach"`
}

type Predicate struct {
	Op      string   `json:"op"`
	Type    string   `json:"type,omitempty"`
	Format  string   `json:"format,omitempty"`
	Pattern string   `json:"pattern,omitempty"`
	Min     *float64 `json:"min,omitempty"`
	Max     *float64 `json:"max,omitempty"`
	Values  []any    `json:"values,omitempty"`
}

type Policy struct {
	Severity string `json:"severity"`
	Action   string `json:"action"`
}

func (c Contract) Validate() error {
	if c.IRVersion != CurrentIRVersion {
		return fmt.Errorf("unsupported ir_version %q, expected %q", c.IRVersion, CurrentIRVersion)
	}
	if c.Metadata.ID == "" || c.Metadata.Version == "" {
		return fmt.Errorf("contract id and version are required")
	}
	if c.Input.Kind != "record-stream" {
		return fmt.Errorf("input.kind must be %q", "record-stream")
	}
	if len(c.Rules) == 0 {
		return fmt.Errorf("at least one rule is required")
	}
	seen := make(map[string]bool, len(c.Rules))
	for _, r := range c.Rules {
		if r.ID == "" || seen[r.ID] {
			return fmt.Errorf("rule ids must be present and unique: %q", r.ID)
		}
		seen[r.ID] = true
		if r.Version != "" && r.Version != "1" {
			return fmt.Errorf("rule %q: unsupported version %q", r.ID, r.Version)
		}
		if r.Kind != "record" {
			return fmt.Errorf("rule %q: kind must be %q", r.ID, "record")
		}
		if r.Path == "" || (r.Path != "$" && !strings.HasPrefix(r.Path, "$.")) {
			return fmt.Errorf("rule %q: path must be $ or start with $.", r.ID)
		}
		if r.Predicate.Op == "" {
			return fmt.Errorf("rule %q: predicate.op is required", r.ID)
		}
		switch r.Predicate.Op {
		case "required":
		case "type":
			if r.Predicate.Type != "string" && r.Predicate.Type != "number" && r.Predicate.Type != "integer" && r.Predicate.Type != "boolean" && r.Predicate.Type != "object" && r.Predicate.Type != "array" {
				return fmt.Errorf("rule %q: unsupported type %q", r.ID, r.Predicate.Type)
			}
		case "format":
			if r.Predicate.Format != "email" {
				return fmt.Errorf("rule %q: unsupported format %q", r.ID, r.Predicate.Format)
			}
		case "regex":
			if r.Predicate.Pattern == "" {
				return fmt.Errorf("rule %q: regex pattern is required", r.ID)
			}
			if _, err := regexp.Compile(r.Predicate.Pattern); err != nil {
				return fmt.Errorf("rule %q: invalid regex: %w", r.ID, err)
			}
		case "range":
			if r.Predicate.Min == nil && r.Predicate.Max == nil {
				return fmt.Errorf("rule %q: range requires min or max", r.ID)
			}
			if r.Predicate.Min != nil && r.Predicate.Max != nil && *r.Predicate.Min > *r.Predicate.Max {
				return fmt.Errorf("rule %q: range min cannot exceed max", r.ID)
			}
		case "enum":
			if len(r.Predicate.Values) == 0 {
				return fmt.Errorf("rule %q: enum values are required", r.ID)
			}
		default:
			return fmt.Errorf("rule %q: unsupported predicate %q", r.ID, r.Predicate.Op)
		}
		if r.OnBreach.Severity != "" && r.OnBreach.Severity != "warning" && r.OnBreach.Severity != "error" {
			return fmt.Errorf("rule %q: severity must be warning or error", r.ID)
		}
		if r.OnBreach.Action == "" {
			return fmt.Errorf("rule %q: action is required", r.ID)
		}
		if r.OnBreach.Action != "warn" && r.OnBreach.Action != "reject" && r.OnBreach.Action != "quarantine" && r.OnBreach.Action != "halt" {
			return fmt.Errorf("rule %q: unsupported action %q", r.ID, r.OnBreach.Action)
		}
	}
	return nil
}

func Decode(data []byte) (Contract, error) {
	var c Contract
	decoder := json.NewDecoder(bytes.NewReader(data))
	decoder.DisallowUnknownFields()
	if err := decoder.Decode(&c); err != nil {
		return Contract{}, fmt.Errorf("decode contract: %w", err)
	}
	var extra any
	if err := decoder.Decode(&extra); err != io.EOF {
		if err == nil {
			return Contract{}, fmt.Errorf("decode contract: multiple JSON values")
		}
		return Contract{}, fmt.Errorf("decode contract: trailing data: %w", err)
	}
	return c.Normalize()
}

// Normalize applies semantic defaults and sorts rules by identity. It returns
// a copy so callers can safely use the input contract for other purposes.
func (c Contract) Normalize() (Contract, error) {
	if err := c.Validate(); err != nil {
		return Contract{}, err
	}
	normalized := c
	normalized.Rules = append([]Rule(nil), c.Rules...)
	for i := range normalized.Rules {
		if normalized.Rules[i].Version == "" {
			normalized.Rules[i].Version = "1"
		}
		if normalized.Rules[i].OnBreach.Severity == "" {
			normalized.Rules[i].OnBreach.Severity = "error"
		}
	}
	sort.Slice(normalized.Rules, func(i, j int) bool {
		return normalized.Rules[i].ID < normalized.Rules[j].ID
	})
	return normalized, nil
}

func CanonicalJSON(c Contract) ([]byte, error) {
	normalized, err := c.Normalize()
	if err != nil {
		return nil, err
	}
	return json.Marshal(normalized)
}

func Digest(c Contract) (string, error) {
	canonical, err := CanonicalJSON(c)
	if err != nil {
		return "", err
	}
	sum := sha256.Sum256(canonical)
	return "sha256:" + fmt.Sprintf("%x", sum), nil
}

File: sdk/python/actually_fine/contract.py
"""Author and compile actually-fine IR 1.0 contracts."""

from __future__ import annotations

from dataclasses import dataclass
import hashlib
import json
import math
import re
from typing import Any, Iterable, Optional

RECORD_OPERATORS = {"required", "not_null", "type", "format", "regex", "range", "enum"}
STREAM_OPERATORS = {"unique", "count"}
OPERATORS = RECORD_OPERATORS | STREAM_OPERATORS
TYPES = {"string", "number", "integer", "boolean", "object", "array"}
ACTIONS = {"warn", "reject", "quarantine", "halt"}
SEVERITIES = {"warning", "error"}


def _number(value: Any) -> Any:
    if isinstance(value, float) and math.isfinite(value) and value.is_integer():
        return int(value)
    if isinstance(value, list):
        return [_number(item) for item in value]
    if isinstance(value, dict):
        return {key: _number(item) for key, item in value.items()}
    return value


@dataclass(frozen=True)
class Predicate:
    op: str
    type: Optional[str] = None
    format: Optional[str] = None
    pattern: Optional[str] = None
    min: Optional[float] = None
    max: Optional[float] = None
    values: Optional[tuple[Any, ...]] = None

    def to_ir(self) -> dict[str, Any]:
        result: dict[str, Any] = {"op": self.op}
        for key in ("type", "format", "pattern", "min", "max"):
            value = getattr(self, key)
            if value is not None:
                result[key] = value
        if self.values is not None:
            result["values"] = list(self.values)
        return result


@dataclass(frozen=True)
class Rule:
    id: str
    path: str
    predicate: Predicate
    action: str
    severity: str = "error"
    kind: Optional[str] = None
    version: str = "1"

    def to_ir(self) -> dict[str, Any]:
        if not self.id:
            raise ValueError("rule id is required")
        expected_kind = "stream" if self.predicate.op in STREAM_OPERATORS else "record"
        if self.predicate.op not in OPERATORS:
            raise ValueError(f"rule {self.id!r}: unsupported predicate {self.predicate.op!r}")
        kind = self.kind or expected_kind
        if kind != expected_kind:
            raise ValueError(f"rule {self.id!r}: {self.predicate.op} must be a {expected_kind} rule")
        if not self.path or (self.path != "$" and not self.path.startswith("$.")):
            raise ValueError(f"rule {self.id!r}: path must be $ or start with $.")
        if self.action not in ACTIONS:
            raise ValueError(f"rule {self.id!r}: unsupported action {self.action!r}")
        if self.severity not in SEVERITIES:
            raise ValueError(f"rule {self.id!r}: severity must be warning or error")
        if self.version != "1":
            raise ValueError(f"rule {self.id!r}: unsupported version {self.version!r}")

        predicate = self.predicate
        if predicate.op == "type" and predicate.type not in TYPES:
            raise ValueError(f"rule {self.id!r}: unsupported type {predicate.type!r}")
        if predicate.op == "format" and predicate.format != "email":
            raise ValueError(f"rule {self.id!r}: unsupported format {predicate.format!r}")
        if predicate.op == "regex":
            if not predicate.pattern:
                raise ValueError(f"rule {self.id!r}: regex pattern is required")
            try:
                re.compile(predicate.pattern)
            except re.error as error:
                raise ValueError(f"rule {self.id!r}: invalid regex: {error}") from error
        if predicate.op in {"range", "count"}:
            if predicate.min is None and predicate.max is None:
                raise ValueError(f"rule {self.id!r}: {predicate.op} requires min or max")
            if predicate.min is not None and predicate.max is not None and predicate.min > predicate.max:
                raise ValueError(f"rule {self.id!r}: {predicate.op} min cannot exceed max")
        if predicate.op == "count" and self.path != "$":
            raise ValueError(f"rule {self.id!r}: count path must be $")
        if predicate.op == "enum" and not predicate.values:
            raise ValueError(f"rule {self.id!r}: enum values are required")

        return {
            "id": self.id,
            "version": self.version,
            "kind": kind,
            "path": self.path,
            "predicate": predicate.to_ir(),
            "on_breach": {"severity": self.severity, "action": self.action},
        }


@dataclass(frozen=True)
class Contract:
    id: str
    version: str
    rules: tuple[Rule, ...]

    def __init__(self, id: str, version: str, rules: Iterable[Rule]):
        object.__setattr__(self, "id", id)
        object.__setattr__(self, "version", version)
        object.__setattr__(self, "rules", tuple(rules))

    def to_ir(self) -> dict[str, Any]:
        if not self.id or not self.version:
            raise ValueError("contract id and version are required")
        if not self.rules:
            raise ValueError("at least one rule is required")
        rules = [rule.to_ir() for rule in self.rules]
        ids = [rule["id"] for rule in rules]
        if len(ids) != len(set(ids)):
            raise ValueError("rule ids must be present and unique")
        rules.sort(key=lambda rule: rule["id"])
        return {
            "ir_version": "1.0",
            "contract": {"id": self.id, "version": self.version},
            "input": {"kind": "record-stream"},
            "rules": rules,
        }

    def canonical_json(self) -> bytes:
        data = _number(self.to_ir())
        encoded = json.dumps(data, ensure_ascii=False, separators=(",", ":"), allow_nan=False)
        # Match encoding/json's escaping for the characters it escapes by default.
        encoded = encoded.replace("&", "\\u0026").replace("<", "\\u003c").replace(">", "\\u003e")
        return encoded.encode("utf-8")

    def digest(self) -> str:
        return "sha256:" + hashlib.sha256(self.canonical_json()).hexdigest()
